new-log-viewer/src/services/decoders/JsonlDecoder/index.ts

```diff
@@ -0,0 +1,242 @@
+
+    #dataArray: Nullable<Uint8Array>;
+
+    #logLevelKey: string;
+
+    #timestampKey: string;
+
+    #logEvents: JsonLogEvent[] = [];
+
+    #filteredLogEventMap: FilteredLogEventMap = null;
+
+    #invalidLogEventIdxToRawLine: Map<number, string> = new Map();
+
+    #formatter: Formatter;
+
+    /**
+     * @param dataArray
+     * @param decoderOptions
+     */
+    constructor (dataArray: Uint8Array, decoderOptions: JsonlDecoderOptionsType) {
+        this.#dataArray = dataArray;
+        this.#logLevelKey = decoderOptions.logLevelKey;
+        this.#timestampKey = decoderOptions.timestampKey;
+        this.#formatter = new LogbackFormatter({formatString: decoderOptions.formatString});
+    }
+
+    getEstimatedNumEvents (): number {
+        return this.#logEvents.length;
+    }
+
+    getFilteredLogEventMap (): FilteredLogEventMap {
+        return this.#filteredLogEventMap;
+    }
+
+    setLogLevelFilter (logLevelFilter: LogLevelFilter): boolean {
+        this.#filterLogs(logLevelFilter);
+
+        return true;
+    }
+
+    build (): LogEventCount {
+        this.#deserialize();
+
+        const numInvalidEvents = Array.from(this.#invalidLogEventIdxToRawLine.keys()).length;
+
+        return {
+            numValidEvents: this.#logEvents.length - numInvalidEvents,
+            numInvalidEvents: numInvalidEvents,
+        };
+    }
+
+    setFormatterOptions (options: JsonlDecoderOptionsType): boolean {
+        this.#formatter = new LogbackFormatter({formatString: options.formatString});
+
+        return true;
+    }
+
+    decodeRange (
+        beginIdx: number,
+        endIdx: number,
+        useFilter: boolean,
+    ): Nullable<DecodeResultType[]> {
+        if (useFilter && null === this.#filteredLogEventMap) {
+            return null;
+        }
+
+        // Prevents typescript potential null warning.
+        const filteredLogEventIndices: number[] = this.#filteredLogEventMap as number[];
+
+        const length: number = useFilter ?
+            filteredLogEventIndices.length :
+            this.#logEvents.length;
+
+        if (0 > beginIdx || length < endIdx) {
+            return null;
+        }
+
+        const results: DecodeResultType[] = [];
+        for (let i = beginIdx; i < endIdx; i++) {
+            // Explicit cast since typescript thinks `#filteredLogEventIndices[filteredLogEventIdx]`
+            // can be undefined, but it shouldn't be since we performed a bounds check at the
+            // beginning of the method.
+            const logEventIdx: number = useFilter ?
+                (filteredLogEventIndices[i] as number) :
+                i;
+
+            results.push(this.#getDecodeResult(logEventIdx));
+        }
+
+        return results;
+    }
+
+    /**
+     * Retrieves log event using index then decodes into `DecodeResultType`.
+     *
+     * @param logEventIdx
+     * @return Decoded result.
+     */
+    #getDecodeResult = (logEventIdx: number): DecodeResultType => {
```

Comment:
* How about `decodeLogEvent`?
* Let's move this to the bottom of file for a more logical ordering.